dali/operators/audio/mfcc/mfcc.cu

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/mfcc/mfcc.h"
+#include <vector>
+#include "dali/kernels/signal/dct/dct_gpu.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+namespace detail {
+
+__global__ void CalcLifterKernel(float *coeffs, int64_t start_idx, int64_t target_length,
+                                 float lifter) {
+  float ampl_mult = lifter / 2;
+  float phase_mult = M_PI / lifter;
+  for (int64_t i = start_idx + threadIdx.x; i < target_length; i += blockDim.x)
+    coeffs[i] = 1.0 + ampl_mult * sinf(phase_mult * (i + 1));
+}
+
+template <>
+DLL_PUBLIC  void LifterCoeffs<GPUBackend>::Calculate(int64_t target_length, float lifter,
+                                                     cudaStream_t stream)  {
+  // If different lifter argument, clear previous coefficients
+  if (lifter_ != lifter) {
+    coeffs_.clear();
+    lifter_ = lifter;
+  }
+
+  // 0 means no liftering
+  if (lifter_ == 0.0f)
+    return;
+
+  // Calculate remaining coefficients (if necessary)
+  if (static_cast<int64_t>(coeffs_.size()) < target_length) {
+    int start_idx = coeffs_.size();
+    int added_length = target_length - start_idx;
+    coeffs_.resize(target_length, stream);
+    int threads = std::min(added_length, 256);
+    CalcLifterKernel<<<1, threads, 0, stream>>>(coeffs_.data(), start_idx, target_length, lifter);
+  }
+}
+
+}  // namespace detail
+
+template<>
+bool MFCC<GPUBackend>::SetupImpl(std::vector<OutputDesc> &output_desc,
+                                 const workspace_t<GPUBackend> &ws) {
+  GetArguments(ws);
+  auto &input = ws.InputRef<GPUBackend>(0);
+  TYPE_SWITCH(input.type().id(), type2id, T, MFCC_SUPPORTED_TYPES, (
```

Comment:
There's a lot going on inside. I'd extract it to SetupTyped - this would give superior compiler diagnostics and precise run-time error traceback.

Reply:
Added statically typed `detail::SetupKernel`